### PR TITLE
fd: add module

### PR DESCRIPTION
**Pull request by uncenter**
### Description

`fd` is a simple, fast and user-friendly alternative to `find`.

This module includes support for the [`.fdignore`/`$XDG_CONFIG_HOME/fd/ignore`](https://github.com/sharkdp/fd#excluding-specific-files-or-directories) global ignore file (`ignores`), the search hidden files and directories `--hidden` option (`hidden`), and `extraOptions` for additional CLI flags.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

N/A

### Additional comments

I've refrained from adding myself to `modules/lib/maintainers.nix` as I have an upstream PR that adds myself as a maintainer in nixpkgs: https://github.com/NixOS/nixpkgs/pull/303399/files#diff-c1fcec3afa1da96245de1d7954ca721c979a5f25aee0e8b59d59d3d21e31f8fa.

### CODE REVIEW

**Comment by rycee** [comment]
Looks good. Thanks for the contribution! I've merged the cargo-shear PR and this can be merged once that PR moves to NixOS unstable (https://nixpk.gs/pr-tracker.html?pr=303399).

File: modules/programs/fd.nix
Comment (loicreynier) [inline modules/programs/fd.nix:76]:
```suggestion
  config = let
    cfg = config.programs.fd;

    args = escapeShellArgs (optional cfg.hidden "--hidden" ++ cfg.extraOptions);

    ignores = mkIf (cfg.ignores != [ ]) {
      text = concatStringsSep "\n" cfg.ignores + "\n";
    };

    optionsAlias = { fd = "fd ${args}"; };
  in mkIf cfg.enable (mkMerge [
    {
      home.packages = [ cfg.package ];

      programs.bash.shellAliases = optionsAlias;

      programs.zsh.shellAliases = optionsAlias;

      programs.fish.shellAliases = optionsAlias;

      programs.ion.shellAliases = optionsAlias;

      programs.nushell.shellAliases = optionsAlias;
    }
    (mkIf (!config.home.preferXdgDirectories) {
      home.file.".fdignore".text = ignores;
    })
    (mkIf (!config.home.preferXdgDirectories) {
      xdg.configFile."fd/ignore" = ignores;
    })
  ]);
```

Comment (loicreynier) [reply modules/programs/fd.nix:76]: We could add something like this to choose between `~/.fdignore`  and `$XDG_CONFIG_HOME/fd/ignore` using `config.home.preferXdgDirectories`.

Comment (uncenter) [reply modules/programs/fd.nix:76]: From `fd`'s man page:

> ### `XDG_CONFIG_HOME`, `HOME`
> Used  to  locate the global ignore file. If `XDG_CONFIG_HOME` is set, use `$XDG_CONFIG_HOME/fd/ignore`.  Otherwise, use `$HOME/.config/fd/ignore`.

I'm fine with that I guess but we should write to `home.file.".config/fd/ignore"` instead?

Comment (rycee) [reply modules/programs/fd.nix:76]: I guess you could have
``` nix
let configPath = if config.home.preferXdgDirectories then "${config.xdg.configHome}/fd/ignore" else ".fdignore";
```

and then do

``` nix
home.file.${configPath}.text = ignores;
```

Comment (loicreynier) [reply modules/programs/fd.nix:76]:
```suggestion
  config = let
    cfg = config.programs.fd;

    args = escapeShellArgs (optional cfg.hidden "--hidden" ++ cfg.extraOptions);

    optionsAlias = { fd = "fd ${args}"; };

    ignores = mkIf (cfg.ignores != [ ]) {
      text = concatStringsSep "\n" cfg.ignores + "\n";
    };

    configPath = if config.home.preferXdgDirectories then
      "${config.xdg.configHome}/fd/ignore"
    else
      ".fdignore";
  in mkIf cfg.enable {
    home.packages = [ cfg.package ];

    programs.bash.shellAliases = optionsAlias;

    programs.zsh.shellAliases = optionsAlias;

    programs.fish.shellAliases = optionsAlias;

    programs.ion.shellAliases = optionsAlias;

    programs.nushell.shellAliases = optionsAlias;

    home.file.${configPath}.text = ignores;
  };
}
```

This indeed looks cleaner.

But if `fd` requires the `XDG_CONFIG_HOME` variable to look for `~/.config/fd/ignores` I guess we should also check for `config.xdg.enable`? Would not it be better to always enable `config.xdg.enable` when setting `home.preferXdgDirectories = true`?

File: modules/programs/fd.nix
Comment (ambroisie) [inline modules/programs/fd.nix:25]: Not enabled by default?

File: modules/programs/fd.nix
Comment (ambroisie) [reply modules/programs/fd.nix:76]: > Used to locate the global ignore file. If XDG_CONFIG_HOME is set, use $XDG_CONFIG_HOME/fd/ignore. Otherwise, use $HOME/.config/fd/ignore.

That's just following the XDG spec no? So always storing it using `xdg.configFile` should just work.

IMO, `preferXdgDirectories` should be avoided in most cases, and used for programs that have no understanding of XDG at all and need e.g: an alias to change the config location.

Anyways, I don't even see `~/.fdignore` mentionned in the manual, just `~/.config/fd/ignore` (the default `XDG_CONFIG_HOME` location).

Comment (uncenter) [reply modules/programs/fd.nix:76]: > Anyways, I don't even see `~/.fdignore` mentionned in the manual, just `~/.config/fd/ignore` (the default `XDG_CONFIG_HOME` location).

That's what I've been trying to say! 😅

Comment (isabelroses) [reply modules/programs/fd.nix:76]: The [documentation](https://github.com/sharkdp/fd#excluding-specific-files-or-directories) only mentions the xdg config home location. So I think this comment can be safely ignored.

Comment (loicreynier) [reply modules/programs/fd.nix:78]: Oh yes indeed! I misunderstood your comment @uncenter, ignore my comments.

**Review by isabelroses — APPROVED** [review]

**Comment by uncenter** [comment]
> Looks good. Thanks for the contribution! I've merged the cargo-shear PR and this can be merged once that PR moves to NixOS unstable ([nixpk.gs/pr-tracker.html?pr=303399](https://nixpk.gs/pr-tracker.html?pr=303399)).

Based on that tracker link it looks like that cargo-shear PR made it to unstable! Anything else I need to do before merging?

File: modules/programs/fd.nix
Comment (isabelroses) [inline modules/programs/fd.nix:51]: Might be better to do something more like ripgrep
https://github.com/nix-community/home-manager/blob/fa8c16e2452bf092ac76f09ee1fb1e9f7d0796e7/modules/programs/ripgrep.nix#L17-L28

Comment (uncenter) [reply modules/programs/fd.nix:51]: I'm going with what the `eza` module did, by having one of the rather important options separate.

File: modules/programs/fd.nix
Comment (rycee) [inline modules/programs/fd.nix:27]: Just noticed that none of these options seem to actually do anything. Perhaps the intent is to use `mkIf` for the various `shellAliases` options?

Comment (uncenter) [reply modules/programs/fd.nix:27]: Oops! Thanks for pointing that out - can't believe I didn't notice. I originally based this off of the `eza` module, but now looking at that one it seems that `eza ${args}` is always aliased no matter if `enable___Integration` is enabled, and the `enable___Integration` options are used to alias things like `ls` and `ll` and so on. I think we can just remove all of those options in that case?

Comment (rycee) [reply modules/programs/fd.nix:27]: Yeah, can remove them.

**Comment by uncenter** [comment]
Ah sorry I didn't know I was supposed to force push in this repository too!

**Comment by rycee** [comment]
No worries. Merged to master now, thanks! 🙂

**Pull request merged**
